dev/src/bulk-writer.ts

```diff
@@ -0,0 +1,497 @@
+   * the Promise waits for all previously issued writes, but it does not wait
+   * for writes that were added after the method is called. If you want to wait
+   * for additional writes, call `waitForPendingWrites()` again.
+   *
+   * @return {Promise<void>} A promise that resolves when there are no more
+   * pending writes.
+   *
+   * @example
+   * let bulkWriter = firestore.bulkWriter();
+   *
+   * bulkWriter.create(documentRef, {foo: 'bar'});
+   * bulkWriter.update(documentRef2, {foo: 'bar'});
+   * bulkWriter.delete(documentRef3);
+   * await waitForPendingWrites().then(() => {
+   *   console.log('Successfully executed all writes');
+   * });
+   */
+  async waitForPendingWrites(): Promise<void> {
+    const trackedBatches = this.batchQueue;
+    const writePromises = trackedBatches.map(batch => batch.awaitBatch());
+    this.sendBatches();
+    await Promise.all(writePromises);
+  }
+
+  /**
+   * Return the first eligible batch that can hold a write to the provided
+   * reference, or creates one if no eligible batches are found.
+   */
+  private getEligibleBatch(ref: DocumentReference): BulkCommitBatch {
+    let batch = null;
+    let logWarning = this.batchQueue.length > 0;
+    for (const bulkBatch of this.batchQueue) {
+      if (!bulkBatch.refsInBatch.has(ref)) {
+        logWarning = false;
+      }
+      if (bulkBatch.canAddDoc(ref)) {
+        batch = bulkBatch;
+      }
+    }
+
+    if (batch === null) {
+      if (logWarning) {
+        console.warn(
+          '[BulkWriter]',
+          'Writing to the same document multiple times will slow down BulkWriter. ' +
+            'Write to unique documents in order to maximize throughput.'
+        );
+      }
+      return this.createNewBatch();
+    } else {
+      return batch;
+    }
+  }
+
+  private createNewBatch(): BulkCommitBatch {
+    const newBatch = new BulkCommitBatch(this.getId(), this.firestore.batch());
+
+    this.batchQueue.push(newBatch);
+    return newBatch;
+  }
+
+  private getId(): number {
+    const id = this.currentId;
+    if (this.currentId === Number.MAX_SAFE_INTEGER) {
+      this.currentId = 0;
+    } else {
+      this.currentId++;
+    }
+    return id;
+  }
+
+  /**
+   * Attempts to send batches starting from the front of the BatchQueue until a
+   * batch cannot be sent.
+   *
+   * After a batch is complete, try sending batches again.
+   */
+  private sendBatches(): void {
+    let counter = 0;
+    for (const batch of this.batchQueue) {
+      // Only send up until the first batch that cannot be sent.
+      if (!this.isBatchSendable(batch)) {
+        break;
+      }
+
+      this.refsInFlight = new Set([...this.refsInFlight, ...batch.refsInBatch]);
+
+      batch.bulkCommit().then(async results => {
+        batch.resolveResults(results);
+
+        // Remove references that are no longer in flight.
+        this.refsInFlight = new Set(
+          [...this.refsInFlight].filter(ref => !batch.refsInBatch.has(ref))
+        );
```

Comment:
See my comment from above. This is the same situation but for removal.  As your comment states you're ultimately removing the refs from the batch from those that are in flight. You can express that more literally through something like `batch.refsInBatch.forEach(b => this.refsInFlight.remove(b))`.

Reply:
done.